### DISCUSSION TITLE
Generic bit_bang in C. I fear we need volatile

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1521] [Q&A]
I am trying to compile some generic C code ideally under several cross-compilers.

My tentative code looks like:
```
void CLICK(uint16_t duration, uint8_t period) 
{
    uint16_t i;
    uint8_t j;
    
    for(i=0;i<duration;++i)
    {
        _click();
        for(j=0;j<period;++j){};
    }
}
```

where `_click()` is a macro that should just do `PEEK(BIT_BANG_ADDR)`.

This does not work as long as I compile it with CC65 and I use `-O`.
Even `__asm__("LDA $C030")` is "optimized away" by  `-O`.

What work is either disabling the optimizer or using an extra variable just to trick the optimizer into not optimizing the code
```
    #define _click() \
    do \
    { \
       uint8_t k; \
       \
       k = PEEK(BIT_BANG_ADDR); \
    } while(0)
```
This is ugly because I am not sure that the next optimization rules won't remove the code anyway.

What is the cleanest and most general solution?

**Comment by oliverschmidt** [comment]
The only, single clean and general solution with cc65 is to call a function. Either an asm function or a C function compiled without optimizations.

**Reply by Fabrizio-Caruso** [reply]
Do you mean to compile just that function without optimization and then link it with the rest?
I hope some sort of volatile support is implemented some day.

**Comment by oliverschmidt** [comment]
No, I mean to put the `PEEK(BIT_BANG_ADDR);` into a separate function.

**Reply by Fabrizio-Caruso** [reply]
Yes, ok but I would like to have -O applied to the rest of the files.

**Reply by oliverschmidt** [reply]
Then use https://cc65.github.io/doc/cc65.html#ss7.11

**Comment by greg-king5** [comment]
Do what Oliver suggested, but go one step further.  Return the peeked value.
```c
static uint8_t _click(void)
{
    return PEEK(BIT_BANG_ADDR);
}
```
The optimizer will think that the value will be used later; therefore, it won't be dropped.  `CLICK()` doesn't use the value, but the optimizer won't know that.

You'll be able to use `-O` on all of the C code.

**Comment by mrdudz** [comment]
i am a bit worried that noone is concerned about
```
for(j=0;j<period;++j){};
```

**Reply by oliverschmidt** [reply]
At least I was not concerned about it as I didn't look at it. The issue at hand was that the speaker wasn't toggled. I only looked at that and commented on that.
